### ISSUE TITLE
Multi-Level-Namespaces

### ISSUE THREAD

**Issue by GWRon** [bmx-ng/bmx-ng#77]
As a bit discussed here and there (eg. in https://github.com/bmx-ng/brl.mod/issues/116) I thought it is better to create a new issue for this idea.

The idea is to allow multiple levels of namespaces:
brl.math.vector
brl.string.stringbuilder
brl.string.lowerstring

While it looks that BMK does not do special stuff for "nested modules" (it utilizes Brl.MaxUtil for some stuff) it seems that BCC requires attention:

bmk.git/config.bmx:
```BlitzMax
Function ModuleHeaderFromIdent:String(ident:String, includePath:Int = False)
	Local ns:String = ident[..ident.find(".")]
	Local name:String = ident[ident.find(".") + 1..]
	
	Local file:String = name + ".bmx" + FileMung() + ".h"
	
	If includePath Then
		file = ns + ".mod/" + name + ".mod/.bmx/" + file
	End If
	
	Return file
End Function
```
So "ns" might be "namespace" - it only reads the "brl" in "brl.math.vector".
"name" is then of course "math.vector".

**Comment by GWRon** [comment]
After this change I was able to have this module:

mod/test.mod/parent.mod/child.mod/child.bmx:
```BlitzMax
SuperStrict

Module test.parent.child

ModuleInfo "Version: 1.0"


Type TChild
	Field i:Int
	
	Method GetI:Int()
		DebugStop
		Return i
	End Method
End Type
```

And a sample file:
```BlitzMax
SuperStrict
Framework Brl.StandardIO

Import test.parent.child

Local c:TChild = New TChild
c.i = 10
Print c.getI()
```
Debugger correctly displayed the method in the child.bmx file.


Surely I've missed something so I keep it here for discussion before even trying to do a PR.

**Comment by GWRon** [comment]
Importing with wildcard: "test.parent.*" should be possible by manipulating "modimports" within BMK (bmk_make.bmx). So if it finds an asterisk it should do a directory scan and import all modules it finds in there.
Yet I think we should be able to even skip the wildcard (except for a "filter").

For each imported module it must be made sure to recursively import all contained modules.
Even if we do not have 4 levels now (and we want to avoid "complexity" now) we assume this structure as example:

mod/brl.mod/math.mod/vectors.mod/vec2d.mod/vec2d.bmx
mod/brl.mod/math.mod/vectors.mod/vec3d.mod/vec3d.bmx
mod/brl.mod/math.mod/boundingbox.mod/boundingbox.bmx

Importing "brl.math" should include boundingbox and both vector modules.

If one really wants wildcards (feels dirty as you never know what is added else) you could check the "module name" and alter your directory traversal to contain filters:
Import brl.math.v*
would then only import modules within "math" starting with "v".
But as said I think this should be avoided - so we should not need "wildcard" (in the sense of strings with an asterisk) support.

**Comment by GWRon** [comment]
@woollybah 
Where are module-names/paths else used in BCC/BMK? 

@HurryStarfish 
What kind of import-statements do you desire to support?

**Comment by GWRon** [comment]
Module:
```BlitzMax
SuperStrict

Module test.parent.child
ModuleInfo "Version: 1.0"

Import Brl.StandardIO


Type TChild
	Field i:Int
	
	Method GetI:Int()
		DebugStop
		Return i
	End Method
End Type


Function ChildFunction:Int()
	Return 33
End Function


'function with same name as in other modules
Function Cls()
	Print "Child.mod Cls()"
End Function
```

sample:
```BlitzMax
SuperStrict
Framework Brl.StandardIO
Import test.parent.child

Local c:TChild = New TChild
c.i = 10
Print c.getI()

Print "test.parent.ChildFunction(): " + test.parent.child.ChildFunction()

test.parent.child.Cls()
```

output: `Compile Error: Identifier 'parent' not found.`. So for this lookup it needs an additional change in BCC/BMK.

**Comment by woollybah** [comment]
In the scheme of things there are a lot more important things to be working on that this... which, invariably, is going to involve a lot of work to get it right.

**Comment by GWRon** [comment]
This is why I at least try to fiddle around to see what is affected and needs changes. Should not harm or distract you too much.

Also I am pretty limited in my knowledge of other (even more complex) things on our ToDo. Seems it's better to do some documentation then...

**Comment by HurryStarfish** [comment]
> If one really wants wildcards (feels dirty as you never know what is added else - which of course happens with "brl.math" too) you could check the "module name" and alter your directory traversal to contain filters:
> Import brl.math.v*
> would then only import modules within "math" starting with "v".
> But as said I think this should be avoided - so we should not need "wildcard" (in the sense of strings with an asterisk) support.

Yeah, I also think this isn't the best idea. When I suggested a `Import BRL.Math.*` syntax in the other thread, I only meant to support `.*`, meaning "import every module in the subfolders/-namespaces of this". Not things like `.v*` to filter by name.

**Comment by HurryStarfish** [comment]
> In the scheme of things there are a lot more important things to be working on than this... which, invariably, is going to involve a lot of work to get it right.

Hmm, what do you think would be the main challenge in implementing this? I imagine that name mangling/resolution would require a bunch of work, to be able to deal with the possible name collisions that may arise in the BlitzMax code/C code/.i files. Is that the main issue, or are there more problems I haven't thought of yet?
